velox/duckdb/conversion/DuckWrapper.cpp

```diff
@@ -267,6 +267,88 @@ VectorPtr convertDecimalToDouble(
   return result;
 }
 
+template <class OP, typename I>
+VectorPtr convertDuckToVeloxDecimal(
+    ::duckdb::Vector& duckVector,
+    TypePtr veloxType,
+    int32_t size,
+    memory::MemoryPool* pool,
+    uint8_t* validity = nullptr) {
+  auto vectorType = duckVector.GetVectorType();
+  auto internalType = duckVector.GetType().InternalType();
+  switch (vectorType) {
+    case ::duckdb::VectorType::FLAT_VECTOR: {
+      VectorPtr result;
+      auto& duckValidity = ::duckdb::FlatVector::Validity(duckVector);
+      auto* duckData = ::duckdb::FlatVector::GetData<I>(duckVector);
+      result = BaseVector::create(veloxType, size, pool);
+      auto flatResult = result->as<FlatVector<typename OP::VELOX_TYPE>>();
+      if (internalType == PhysicalType::INT16 ||
+          internalType == PhysicalType::INT32) {
+        // Cannot re-use duckdb buffers. Create a copy of the vector.
+        for (auto i = 0; i < size; i++) {
+          if (duckValidity.RowIsValid(i) &&
+              (!validity || bits::isBitSet(validity, i))) {
+            flatResult->set(i, OP::toVelox(duckData[i]));
+          }
+        }
+        if (!duckValidity.AllValid()) {
+          auto rawNulls = flatResult->mutableRawNulls();
+          memcpy(rawNulls, duckValidity.GetData(), bits::nbytes(size));
+        }
+      } else if (
+          internalType == PhysicalType::INT64 ||
+          internalType == PhysicalType::INT128) {
+        // Re-use duckdb buffers.
+        auto valuesView = BufferView<DuckDBBufferReleaser>::create(
+            reinterpret_cast<const uint8_t*>(duckData),
+            size * sizeof(typename OP::VELOX_TYPE),
+            DuckDBBufferReleaser(duckVector.GetBuffer()));
+        BufferPtr nullsView(nullptr);
+        if (!duckValidity.AllValid()) {
+          nullsView = BufferView<DuckDBValidityReleaser>::create(
+              reinterpret_cast<const uint8_t*>(duckValidity.GetData()),
+              bits::nbytes(size),
+              DuckDBValidityReleaser(duckValidity));
+        }
+        result = std::make_shared<FlatVector<typename OP::VELOX_TYPE>>(
+            pool,
+            veloxType,
+            nullsView,
+            size,
+            valuesView,
+            std::vector<BufferPtr>());
+      } else {
+        VELOX_UNSUPPORTED(
+            "Unsupported DuckDB LogicalType:{} to ShortDecimal "
+            "conversion",
+            duckVector.GetType().ToString());
+      }
+      return result;
+    }
+    case ::duckdb::VectorType::DICTIONARY_VECTOR: {
```

Comment:
Why do we handle dictionary here? Shouldn't dictionary be processed in a type-agnostic way?

Reply:
I am following the general case of converting Duck vectors to Velox vectors here which as well handles Dictionary with in it.
```
template <class OP>
VectorPtr convert(
    ::duckdb::Vector& duckVector,
    const TypePtr& veloxType,
    size_t size,
    memory::MemoryPool* pool,
    uint8_t* validity = nullptr) {
```
The step before convert() calls is toVeloxVector which doesn't do any thing specific to a vector type.

Reply:
The logic for converting dictionary vectors is the same for all types, hence, it makes sense to refactor it into a single type-agnostic place rather than duplicate.